### PR TITLE
feat: path completion take 2

### PR DESCRIPTION
**Pull request by sunjon**

### CODE REVIEW

File: lua/complementree/utils.lua [hunk 1]
```diff
@@ -9,4 +13,73 @@ function M.cword(complete_item)
...
+      if not is_ignored_path(path) then
+        sub_dirs[#sub_dirs + 1] = path
+      end
+    elseif ftype == 'file' then
+      if not is_ignored_path(path) then
+        if opts.match_patterns ~= {} and not is_matched_filetype(path) then
+          goto continue
+        end
+        list[#list + 1] = ('%s%s%s'):format(root_dir, os_sep, path)
+        ::continue::
```
Comment (vigoux) [inline lua/complementree/utils.lua]: Why not inverting it ?
```suggestion
        if opts.match_patterns == {} or is_matched_filetype(path) then
                list[#list + 1] = ('%s%s%s'):format(root_dir, os_sep, path)
        end
```

It avoids some complexities.

**Review by vigoux — COMMENTED** [review]
Looks nice

File: lua/complementree/sources.lua [hunk 2]
```diff
@@ -166,6 +165,67 @@ local function lsp_completedone(completed_item)
...
+    for _, i in ipairs(items) do
+      display_path = config.relative_paths and relpath(i.path, i.root_dir) or i.path
+
+      matches[#matches + 1] = {
+        word = i.path,
+        abbr = display_path,
+        kind = '[path]',
+        icase = 1,
+        dup = 1,
+        empty = 1,
```
Comment (vigoux) [inline lua/complementree/sources.lua:358]: Could you put a little `equals = 1` here too ? We need it to tell vim we handle the comparison.
```suggestion
        empty = 1,
        equals = 1,
```

File: lua/complementree/utils.lua [hunk 3]
```diff
@@ -9,4 +13,71 @@ function M.cword(complete_item)
...
+  return path
+end
+
+local scandir_default_opts = {
+  ignore_hidden = true,
+  max_depth = 8,
+  match_patterns = {},
+}
+
+M.scan_dir = function(root_dir, opts, list, depth)
```
Comment (vigoux) [inline lua/complementree/utils.lua]: I tend to like it the other way:
```suggestion
function M.scan_dir(root_dir, opts, list, depth)
```

File: lua/complementree/utils.lua [hunk 4]
```diff
@@ -9,4 +13,71 @@ function M.cword(complete_item)
   return (complete_item.word or complete_item.abbr)
 end
 
+M.make_relative_path = function(path, root)
```
Comment (vigoux) [inline lua/complementree/utils.lua]: Same as below here.

File: lua/complementree/utils.lua [hunk 5]
```diff
@@ -9,4 +13,71 @@ function M.cword(complete_item)
...
+  local sub_dirs = {}
+  for path, ftype in iter do
+    if ftype == 'directory' then
+      if not is_ignored_path(path) then
+        sub_dirs[#sub_dirs + 1] = path
+      end
+    elseif ftype == 'file' then
+      if not is_ignored_path(path) then
+        if vim.tbl_isempty(opts.match_patterns) or is_matched_filetype(path) then
+          list[#list + 1] = ('%s%s%s'):format(root_dir, os_sep, path)
```
Comment (vigoux) [inline lua/complementree/utils.lua]: I'd rather like all occurences of that to become:
```suggestion
          list[#list + 1] = string.format('%s%s%s', root_dir, os_sep, path)
```

**Comment by sunjon** [comment]
```lua
  local fp_matcher = sources.filepath_matches {
    max_depth = 8,
    ignore_hidden = true,
    relative_paths = true,
    root_dirs = { "/home/sunjon" },
    match_patterns = { "%.png$", "%.jpg$" }
  }
```
I've put the options here as a reminder to add them to the README when the PR is complete.

File: lua/complementree/sources.lua [hunk 6]
```diff
@@ -166,6 +165,111 @@ local function lsp_completedone(completed_item)
...
+            return full_path
+          end
+        else
+          return full_path
+        end
+      end
+    end
+  end
+end
+
```
Comment (vigoux) [inline lua/complementree/sources.lua]: Why not moving all this inside the filepath_matches function ? This way I guess the the function will be a bit more efficient ass all the options are going to be constants ?

File: lua/complementree/sources.lua [hunk 7]
```diff
@@ -166,6 +165,111 @@ local function lsp_completedone(completed_item)
...
+    match_patterns = opts.match_patterns or {},
+  }
+
+  return cached('filepath', function(line_to_cursor, _)
+    local pref_start = line_to_cursor:find '%w*$'
+    local prefix = line_to_cursor:sub(pref_start)
+
+    -- local items = {}
+    local items = {}
+    for path in iter_files(opts) do
```
Comment (vigoux) [inline lua/complementree/sources.lua]: Here again, the `opts` table is constant across the completion attempts, thus we could move that out of the function.

File: lua/complementree/sources.lua [hunk 8]
```diff
@@ -166,6 +165,111 @@ local function lsp_completedone(completed_item)
...
+  local config = {
+    max_depth = opts.max_depth or 8,
+    show_hidden = opts.show_hidden or true,
+    relative_paths = opts.relative_paths or false,
+    root_dirs = opts.root_dirs,
+    match_patterns = opts.match_patterns or {},
+  }
+
+  return cached('filepath', function(line_to_cursor, _)
+    local pref_start = line_to_cursor:find '%w*$'
```
Comment (vigoux) [inline lua/complementree/sources.lua]: For files I am not sure this is actually the prefix you want, you might want to include `-`, `_`, `.`, and the path separator for this platform.

File: lua/complementree/utils.lua (hunk 5 above)
Comment (sunjon) [reply lua/complementree/utils.lua]: That function was removed, but I'll bear it in mind in future.
Any particular reason you prefer that style?

Comment (vigoux) [reply lua/complementree/utils.lua]: I find it clearer.

File: lua/complementree/utils.lua [hunk 9]
```diff
@@ -9,4 +9,12 @@ function M.cword(complete_item)
   return (complete_item.word or complete_item.abbr)
 end
 
+function M.make_relative_path(path, root)
```
Comment (vigoux) [inline lua/complementree/utils.lua:12]: This assumes that `root` is a prefix of `path`. Could you add a check for that using `vim.startswith` and error out if it is not the case ?

File: lua/complementree/sources.lua [hunk 10]
```diff
@@ -166,6 +165,112 @@ local function lsp_completedone(completed_item)
...
+          end
+        end
+      end
+    end
+  end
+
+  return cached('filepath', function(line_to_cursor, _)
+    -- local pref_start = line_to_cursor:find '%w*$'
+    local pref_start = line_to_cursor:find(os_path)
+    print("A:" .. pref_start)
```
Comment (vigoux) [inline lua/complementree/sources.lua]:
```suggestion
```

File: lua/complementree/sources.lua [hunk 11]
```diff
@@ -166,6 +165,112 @@ local function lsp_completedone(completed_item)
...
+    local prefix = line_to_cursor:sub(pref_start)
+    print("B:" .. prefix)
```
Comment (vigoux) [inline lua/complementree/sources.lua]:
```suggestion
```

File: lua/complementree/sources.lua [hunk 12]
```diff
@@ -166,6 +165,109 @@ local function lsp_completedone(completed_item)
...
 end
 
+--
+
+local os = string.lower(jit.os)
+local is_linux = (os == 'linux' or os == 'osx' or os == 'bsd')
+local os_sep = is_linux and '/' or '\\'
+local os_path = '[' .. os_sep ..'%w+%-%.%_]*$'
+
+M.filepath_matches = function(opts)
```
Comment (vigoux) [inline lua/complementree/sources.lua]: Could you switch to the other style:
```suggestion
function M.filepath_matches(opts)
```

File: lua/complementree/sources.lua [hunk 13]
```diff
@@ -166,6 +165,109 @@ local function lsp_completedone(completed_item)
...
+
+    local cwd = vim.fn.getcwd()
+    local display_path
+    local matches = {}
+    for _, path in ipairs(items) do
+      display_path = config.relative_paths and relpath(path, cwd) or path
+
+      matches[#matches + 1] = {
+        word = path,
+        abbr = display_path,
```
Comment (vigoux) [inline lua/complementree/sources.lua]: So you will display the relative path but complete using the full path ? isn't that a bit nasty :wink:

File: lua/complementree/sources.lua [hunk 14]
```diff
@@ -166,6 +165,109 @@ local function lsp_completedone(completed_item)
...
+    -- local items = {}
+    local items = {}
+    for path in iter_files() do
+      items[#items + 1] = path
+    end
+
+    local cwd = vim.fn.getcwd()
+    local display_path
+    local matches = {}
+    for _, path in ipairs(items) do
```
Comment (vigoux) [inline lua/complementree/sources.lua]:
```suggestion
    local cwd = vim.fn.getcwd()
    local display_path
    local matches = {}
    for path in iter_files() do
```

**Review by vigoux — COMMENTED** [review]
It looks really good, some minor comments and we're merging this.

**Review by vigoux — APPROVED** [review]

**Pull request merged**